src/V3AstNodes.cpp

```diff
@@ -134,6 +135,21 @@ string AstCCall::selfPointerProtect(bool useSelfForThis) const {
     return VIdProtect::protectWordsIf(sp, protect());
 }
 
+AstNodeCond::AstNodeCond(VNType t, FileLine* fl, AstNodeExpr* condp, AstNodeExpr* thenp,
+                         AstNodeExpr* elsep)
+    : AstNodeTriop{t, fl, condp, thenp, elsep} {
+    UASSERT_OBJ(thenp, this, "No thenp expression");
+    UASSERT_OBJ(elsep, this, "No elsep expression");
+    if (thenp->isClassHandleValue() && elsep->isClassHandleValue()) {
+        // Get the most-deriving class type that both arguments can be casted to.
+        AstNodeDType* const commonClassTypep = V3Width::getCommonClassTypep(thenp, elsep);
+        UASSERT_OBJ(commonClassTypep, this, "No common base class exists");
+        v3Global.rootp()->typeTablep()->addTypesp(commonClassTypep);
```

Comment:
In the current implementation it needs to be added. The test I added fails without it.
It is possible that the condition operator will have a type that no other expression has. Here is an example:
```systemverilog
class A; endclass
class B extends A; endclass
class C extends B; endclass
class D extends B; endclass

module t(input logic cond);
   initial begin
      C c = new;
      D d = new;
      A a = null;
      if (cond) a = c;
      else a = d;
   end
endmodule
```
The condition operation (to which `if/else` will be converted) should have the type `B`, because it is the nearest common type.

If a common class type is a type of one of its sub-expressions, `commonClassTypep` returns a copy of that type. The copying of the type may be unnecessary, but otherwise we would need to iterate through the whole type table to check if the type wasn't added already. And the call of that function in constructor returns a class type only for AstNodeCond nodes created after V3Width (at earlier stages dtypep pointers are usually nullptr).

Reply:
If we accept the copying, it should work on a case with two similar conditions. To be sure, I duplicated statements in the test I added and it passed.

Reply:
I'd prefer not to copy as I think this may result in a lot of new types.  I suppose it will be relatively uncommon that the type needed is not one of the argument dtypes already (given the old code worked in most cases), so instead you could compare the type to the two operands, and only type add it if it doesn't match either (otherwise use the appropriate operand's type.)

Reply:
I am not sure why `V3Width::getCommonClassTypep` returns a copy of the type. When I wrote it, I thought it is needed. I removed the copying.
I also noticed that the type that was found is really always in the type table. Even if no expression has that type, it is in the type table, because AstClassExtends has that dtype.